controllers/routing/routing_delete_resources.go

```diff
@@ -0,0 +1,75 @@
+package routing
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/opendatahub-io/odh-platform/pkg/metadata"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// HandleResourceDeletion handles the removal of dependent resources when the target resource is being deleted.
+func (r *PlatformRoutingReconciler) HandleResourceDeletion(ctx context.Context, sourceRes *unstructured.Unstructured) (ctrl.Result, error) {
+	exportModes, found := extractExportModes(sourceRes)
+	if !found {
+		r.log.Info("No export modes found, skipping deletion logic", "sourceRes", sourceRes)
+
+		return ctrl.Result{}, nil
+	}
+
+	r.log.Info("Handling deletion of dependent resources", "sourceRes", sourceRes)
+
+	for _, exportMode := range exportModes {
+		if err := r.deleteResourcesByLabels(ctx, sourceRes, metadata.ResourceGVKs(exportMode)); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to delete resources for export mode %s: %w", exportMode, err)
+		}
+	}
+
+	return removeFinalizerAndUpdate(ctx, r.Client, sourceRes)
+}
+
+func (r *PlatformRoutingReconciler) deleteResourcesByLabels(ctx context.Context, target *unstructured.Unstructured, gvkList []schema.GroupVersionKind) error {
+	ownerName := target.GetName()
+	ownerKind := target.GetObjectKind().GroupVersionKind().Kind
+	ownerUID := string(target.GetUID())
+
+	labelSelector := client.MatchingLabels{
+		metadata.Labels.OwnerName: ownerName,
+		metadata.Labels.OwnerKind: ownerKind,
+		metadata.Labels.OwnerUID:  ownerUID,
+	}
+
+	for _, gvk := range gvkList {
+		resource := &unstructured.Unstructured{}
+		resource.SetGroupVersionKind(gvk)
+
+		deleteOptions := []client.DeleteAllOfOption{
+			client.InNamespace(r.config.GatewayNamespace),
+			labelSelector,
+		}
```

Comment:
No need to re-instantiate this in the loop, can be moved up next to `labelSelector`. And on that note - perhaps giving `labelSelector` more accurate name will help :)

Reply:
[50cb72b](https://github.com/maistra/odh-platform/pull/39/commits/50cb72b69dbc56c67a394eefc709c9131e5054af)